### PR TITLE
Feature/home page layout

### LINKED ISSUE #26: iOS Feature - Home Page UI
## Feature

- [x] basic layout
  - [x] tab bar
  - [x] navigation bar
- [x] home page layout design
- [x] explore page layout design

# preview
image

### PR DESCRIPTION
**Pull request by Vince-Ku**
Please check out a preview picture in the [issue](https://github.com/No-57/SCSH_3C_iOS/issues/26).

### CODE REVIEW

File: SCSH_3C_iOS/SCSH_3C_iOS/Modules/Explore/View/BoardSectionCollectionViewCell.swift [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+    private func setupButtons() {
+        previousButton.backgroundColor = .black
+        previousButton.alpha = 0
+        previousButton.setTitle("<", for: .normal)
+        previousButton.addTarget(self, action: #selector(previousButtonDidTap), for: .touchUpInside)
+
+        nextButton.backgroundColor = .black
+        nextButton.alpha = 0
+        nextButton.setTitle(">", for: .normal)
+        nextButton.addTarget(self, action: #selector(nextButtonDidTap), for: .touchUpInside)
```
Comment (shadow3x3x3) [inline SCSH_3C_iOS/SCSH_3C_iOS/Modules/Explore/View/BoardSectionCollectionViewCell.swift:107]: IIUC, the setup logic for previous/next button follow a similar pattern. 

Have you thought about wrapping it in a function or method?

File: SCSH_3C_iOS/SCSH_3C_iOS/Modules/Explore/View/BoardSectionCollectionViewCell.swift [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+    
+    func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
+        let cell = collectionView.dequeueReusableCell(withReuseIdentifier: "BoardCollectionViewCell", for: indexPath) as! BoardCollectionViewCell
+
+        // TODO: implement it.
+        if indexPath.item % items.count == 0 {
+            cell.contentView.backgroundColor = .red
+        } else {
+            cell.contentView.backgroundColor = .black
+        }
```
Comment (shadow3x3x3) [inline SCSH_3C_iOS/SCSH_3C_iOS/Modules/Explore/View/BoardSectionCollectionViewCell.swift]: I'm not entirely certain if this is a widely used convention in Swift, but in my view

```swift
cell.contentView.backgroundColor = .black
if indexPath.item % items.count == 0 {
    cell.contentView.backgroundColor = .red
}
```
appears more favorable

File: SCSH_3C_iOS/SCSH_3C_iOS/Modules/Explore/View/ExploreViewController.swift [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+            return cell
+            
+        case SectionName.Explore.rawValue:
+            let cell = collectionView.dequeueReusableCell(withReuseIdentifier: "ExploreSectionCollectionViewCell", for: indexPath) as! ExploreSectionCollectionViewCell
+            cell.contentView.backgroundColor = .blue
+            return cell
+            
+        default:
+            return UICollectionViewCell()
+        }
```
Comment (shadow3x3x3) [inline SCSH_3C_iOS/SCSH_3C_iOS/Modules/Explore/View/ExploreViewController.swift:137]: Do you think the code would be cleaner if eliminate 'return cell' from each case and have a single return statement after the switch?

Or it is impossible to declare `cell` using a generic(interface) type?

File: SCSH_3C_iOS/SCSH_3C_iOS/Modules/Home/View/BodyCollectionViewCell.swift [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+        self.embeddedViewController = embeddedviewController
+
+        // Add the view controller as a child view controller
+        parentViewController.addChild(embeddedviewController)
+
+        // Set the frame of the view controller's view
+        embeddedviewController.view.frame = self.bodyView.bounds
+
+        // Add the view controller's view to the cell's content view
+        self.bodyView.addSubview(embeddedviewController.view)
```
Comment (shadow3x3x3) [inline SCSH_3C_iOS/SCSH_3C_iOS/Modules/Home/View/BodyCollectionViewCell.swift]: I believe that the comments here are not needed. The code appears to be self-explanatory.

There are some comments in other parts of the code that may not be needed as well.

File: SCSH_3C_iOS/SCSH_3C_iOS/Modules/Home/View/HomeViewController.swift [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+    /// Header
+    ///         []  []  []
+    /// Body
+    ///         [][][][][]
+    ///
+    /// When the Body swips `5` blocks, the Header needs to swipe `2` blocks.
+    ///
+    func scrollViewDidScroll(_ scrollView: UIScrollView) {
+        guard scrollView == bodyCollectionView else { return }
+        headerCollectionView.contentOffset.x = scrollView.contentOffset.x * (2 / 5)
```
Comment (shadow3x3x3) [inline SCSH_3C_iOS/SCSH_3C_iOS/Modules/Home/View/HomeViewController.swift]: The numbers 5 and 2 appear to have specific significance in this context and are being consistently used. 
I recommend declaring them as constants for better clarity.

File: SCSH_3C_iOS/SCSH_3C_iOS/Modules/Explore/View/BoardSectionCollectionViewCell.swift (hunk 1 above)
Comment (Vince-Ku) [reply SCSH_3C_iOS/SCSH_3C_iOS/Modules/Explore/View/BoardSectionCollectionViewCell.swift:107]: I used to not over-encapsulate when I implemented `UI` for the first time,
From my experience, **UI changes frequently all the time !!,** I only do so if I've double-confirmed with the UI/UX designer from the beginning, or if the app version is stable.

In this case, it's `NOT` over-encapsulation, but I still prefer not to do it 😅

File: SCSH_3C_iOS/SCSH_3C_iOS/Modules/Explore/View/BoardSectionCollectionViewCell.swift (hunk 2 above)
Comment (Vince-Ku) [reply SCSH_3C_iOS/SCSH_3C_iOS/Modules/Explore/View/BoardSectionCollectionViewCell.swift]: Your version is more concise, I'll update it to yours.

File: SCSH_3C_iOS/SCSH_3C_iOS/Modules/Explore/View/BoardSectionCollectionViewCell.swift (hunk 1 above)
Comment (shadow3x3x3) [reply SCSH_3C_iOS/SCSH_3C_iOS/Modules/Explore/View/BoardSectionCollectionViewCell.swift:107]: It makes sense.

File: SCSH_3C_iOS/SCSH_3C_iOS/Modules/Explore/View/ExploreViewController.swift (hunk 3 above)
Comment (Vince-Ku) [reply SCSH_3C_iOS/SCSH_3C_iOS/Modules/Explore/View/ExploreViewController.swift:137]: > Do you think the code would be cleaner if eliminate 'return cell' from each case and have a single return statement 
> after the switch?

In the single return version, I can't get the casting type at the compile-time. ( Static / Dynamic type issue )

**For example**
```
switch indexPath.section {
        let cell: UICollectionViewCell
            
        case SectionName.Board.rawValue:
            cell = collectionView.dequeueReusableCell(withReuseIdentifier: "BoardSectionCollectionViewCell", for: indexPath) as! BoardSectionCollectionViewCell
           
           /// CATN'T get type `BoardSectionCollectionViewCell` at the compile-time.
           
        default:
            cell = UICollectionViewCell()
        }
           return cell
}
```

File: SCSH_3C_iOS/SCSH_3C_iOS/Modules/Home/View/HomeViewController.swift [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+            
+            // get the actual index for subjects array
+            var index = indexPath.item % viewModel.subjects.count
+            
+            // Since one page contains three header cells, the primary cell should be centered on the page.
+            if index == 0 {
+                index = viewModel.subjects.count - 1
+            } else {
+                index -= 1
+            }
```
Comment (shadow3x3x3) [inline SCSH_3C_iOS/SCSH_3C_iOS/Modules/Home/View/HomeViewController.swift:159]: nit: How about using ternary operator?

```swift
index = (index == 0)
    ? viewModel.subjects.count - 1
    : index - 1
```

File: SCSH_3C_iOS/SCSH_3C_iOS/Modules/Explore/View/ExploreViewController.swift (hunk 3 above)
Comment (shadow3x3x3) [reply SCSH_3C_iOS/SCSH_3C_iOS/Modules/Explore/View/ExploreViewController.swift:137]: I got it.

Comment (Vince-Ku) [reply SCSH_3C_iOS/SCSH_3C_iOS/Modules/Explore/View/ExploreViewController.swift:137]: > Or it is impossible to declare cell using a generic(interface) type?

I'm not quite sure, but I prefer to implement UI in a straightforward way at the beginning.

File: SCSH_3C_iOS/SCSH_3C_iOS/Modules/Home/View/BodyCollectionViewCell.swift (hunk 4 above)
Comment (Vince-Ku) [reply SCSH_3C_iOS/SCSH_3C_iOS/Modules/Home/View/BodyCollectionViewCell.swift]: absolutely agree, I forgot to delete it 😅

File: SCSH_3C_iOS/SCSH_3C_iOS/Modules/Home/View/HomeViewController.swift (hunk 5 above)
Comment (Vince-Ku) [reply SCSH_3C_iOS/SCSH_3C_iOS/Modules/Home/View/HomeViewController.swift]: Agree, I'll update it.

File: SCSH_3C_iOS/SCSH_3C_iOS/Modules/Home/View/HomeViewController.swift (hunk 6 above)
Comment (Vince-Ku) [reply SCSH_3C_iOS/SCSH_3C_iOS/Modules/Home/View/HomeViewController.swift:159]: I rarely used the ternary operator unless the logic was simple. 

In this case, I prefer using `IF-ELSE` control flow.

File: SCSH_3C_iOS/SCSH_3C_iOS/Modules/Explore/View/BoardSectionCollectionViewCell.swift (hunk 2 above)
Comment (Vince-Ku) [reply SCSH_3C_iOS/SCSH_3C_iOS/Modules/Explore/View/BoardSectionCollectionViewCell.swift]: [done](https://github.com/No-57/SCSH_3C_iOS/pull/27/commits/4b084cdbf5568d1ef940c185f9d2dfdbebccee62)

File: SCSH_3C_iOS/SCSH_3C_iOS/Modules/Home/View/BodyCollectionViewCell.swift (hunk 4 above)
Comment (Vince-Ku) [reply SCSH_3C_iOS/SCSH_3C_iOS/Modules/Home/View/BodyCollectionViewCell.swift]: [done](https://github.com/No-57/SCSH_3C_iOS/pull/27/commits/30ae5ec3e8ab3cafaa1fa0be661683db15d23d4e)

File: SCSH_3C_iOS/SCSH_3C_iOS/Modules/Home/View/HomeViewController.swift (hunk 5 above)
Comment (Vince-Ku) [reply SCSH_3C_iOS/SCSH_3C_iOS/Modules/Home/View/HomeViewController.swift]: [done](https://github.com/No-57/SCSH_3C_iOS/pull/27/commits/da17810c430e4915f0b6ae0008e3b491fdc2b755)

**Review by shadow3x3x3 — APPROVED** [review]

**Pull request merged**